staging/src/k8s.io/client-go/tools/events/event_broadcaster.go

```diff
@@ -0,0 +1,302 @@
+	*watch.Broadcaster
+	eventCache    map[eventKey]*v1beta1.Event
+	mu            sync.RWMutex
+	sleepDuration time.Duration
+	sink          EventSink
+}
+
+// NewBroadcaster Creates a new event broadcaster.
+func NewBroadcaster(sink EventSink) EventBroadcaster {
+	return newBroadcaster(sink, defaultSleepDuration)
+}
+
+// NewBroadcasterForTest Creates a new event broadcaster for test purposes.
+func newBroadcaster(sink EventSink, sleepDuration time.Duration) EventBroadcaster {
+	return &eventBroadcasterImpl{
+		Broadcaster:   watch.NewBroadcaster(maxQueuedEvents, watch.DropIfChannelFull),
+		eventCache:    map[eventKey]*v1beta1.Event{},
+		sleepDuration: sleepDuration,
+		sink:          sink,
+	}
+}
+
+func (e *eventBroadcasterImpl) refreshExistingEventSeries() {
+	e.mu.RLock()
+	defer e.mu.RUnlock()
+	for _, event := range e.eventCache {
+
+		if event.Series != nil {
+			//TODO: investigate locking
+			if recordedEvent, retry := recordEvent(e.sink, event); !retry {
+				event = recordedEvent
+			}
+		}
+
+	}
+}
+
+func (e *eventBroadcasterImpl) finishSeries() {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	for isomorphicKey, event := range e.eventCache {
+		eventSerie := event.Series
+		if eventSerie != nil {
+			if eventSerie.LastObservedTime.Time.Add(finishTime).Before(time.Now()) {
+				//TODO: investigate locking
+				if recordedEvent, retry := recordEvent(e.sink, event); !retry {
+					event = recordedEvent
+					delete(e.eventCache, isomorphicKey)
+				}
+			}
+		}
+
+	}
+}
+
+// NewRecorder returns an EventRecorder that records events with the given event source.
+func (e *eventBroadcasterImpl) NewRecorder(scheme *runtime.Scheme, reportingController string) EventRecorder {
+
+	hostname, _ := os.Hostname()
+	reportingInstance := reportingController + "-" + hostname
+
+	return &recorderImpl{scheme, reportingController, reportingInstance, e.Broadcaster, clock.RealClock{}}
+}
+
+func (e *eventBroadcasterImpl) recordToSink(event *v1beta1.Event, clock clock.Clock) {
+	// Make a copy before modification, because there could be multiple listeners.
+	// Events are safe to copy like this.
+	eventCopy := *event
+	event = &eventCopy
+
+	go func() {
+		eventKey := getKey(event)
+		e.mu.RLock()
+		isomorphicEvent, isIsomorphic := e.eventCache[eventKey]
+		e.mu.RUnlock()
+
+		if isIsomorphic {
+
+			if isomorphicEvent.Series != nil {
+				isomorphicEvent.Series.Count++
+				isomorphicEvent.EventTime = metav1.MicroTime{Time: clock.Now()}
+			} else {
+				isomorphicEvent.Series = &v1beta1.EventSeries{
+					Count:            1,
+					LastObservedTime: metav1.MicroTime{Time: clock.Now()},
+				}
+				e.attemptRecording(isomorphicEvent)
+			}
+
+		} else {
+			e.mu.Lock()
+			e.eventCache[eventKey] = event
+			e.mu.Unlock()
+			e.attemptRecording(event)
+		}
+
+	}()
+
+}
+
+func (e *eventBroadcasterImpl) attemptRecording(event *v1beta1.Event) {
+	tries := 0
+
+	for {
+		if recordedEvent, retry := recordEvent(e.sink, event); !retry {
+			event = recordedEvent
```

Comment:
I'm not a fan of overwriting input parameter (especially given the fact we don't seem to take advantage of it anywhere).
If really needed, please return it explicitly (but i don't think you need it).

Reply:
if we don't update the event, any admission controller mutating one of the fields of the isomorphic key on an event could break the logic ?

Reply:
Not sure I'm following - can you please clarify?

Reply:
The idea here is to update re-affect what was returned from the Sink (e.g. kube client). IIUC what is returned to the sink is what gets persisted to etcd (thus passing the admission phase). What happens if a user sets up a mutating webhook on event, the cache will not reflect those mutations.

Reply:
I think I still don't fully understand it  - webhooks, admission etc. is happening on the server (which obviously doesn't have the pointer to it, as it was already sent over network). Am I missing something?

Reply:
ohh - so what you want to achieve is to update the cache based on what was returned from the etcd. That makes sense, but I would really like to make it explicitly, rather than updating input pointer here.

Instead of that, please explicitly return recordedEvent, and when handling the function update the cache.

Reply:
sgtm as a solution